### PR TITLE
Completed dao6

### PR DESCRIPTION
**Pull request by Vedroid**

### CODE REVIEW

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import ua.vedroid.dao3.model.Car;
+import ua.vedroid.dao3.model.Driver;
+import ua.vedroid.dao3.model.Manufacturer;
+import ua.vedroid.dao3.util.ConnectionUtil;
+
+@Dao
+public class CarDaoJdbcImpl implements CarDao {
+    @Override
+    public Car create(Car car) {
+        String query = "INSERT INTO cars(m_id, model) VALUES (?, ?);";
```
Comment (Artificial-Friend) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: I don't know if you're violating checkstyle like that... 
image
`m` is good for shortened variant but as a full name I could be a little bit frustrated about `m` meaning.

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            return Optional.empty();
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Get car with id="
+                    + id + " is failed", ex);
+        }
+    }
+
+    @Override
+    public List<Car> getAll() {
+        String query = "SELECT * FROM cars JOIN manufacturers m on m.id = cars.m_id "
```
Comment (Artificial-Friend) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: I was told that star select is a bad practice. TBH not sure about that but be advised.

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    @Override
+    public Car update(Car car) {
+        String query = "UPDATE cars SET m_id = ?, model = ?"
+                + " WHERE id = ? AND is_deleted = false;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            removeDrivers(connection, car.getId());
+            statement.setLong(1, car.getManufacturer().getId());
```
Comment (Artificial-Friend) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: Are you sure you want to set id manually and mess with database autoincrement system?..

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, driverId);
+            ResultSet resultSet = statement.executeQuery();
+            List<Car> cars = new ArrayList<>();
+            while (resultSet.next()) {
+                cars.add(getCarFromResultSet(connection, resultSet));
+            }
+            return cars;
+        } catch (SQLException ex) {
+            throw new DataProcessingException("", ex);
```
Comment (Artificial-Friend) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: Everything is on fire and nobody knows why.

**Review by Artificial-Friend — APPROVED** [review]

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 2 above)
Comment (dmytro-hryhoruk) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: @Vedroid @Artificial-Friend The thing is that you may add new fields to your table later, so it's better to specify here which exactly field you want to select(even though now its all of them);

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 4 above)
Comment (dmytro-hryhoruk) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: agree

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 1 above)
Comment (dmytro-hryhoruk) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: agree

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 2 above)
Comment (Artificial-Friend) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: Now I get it, thank you!

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+                PreparedStatement statement =
+                        connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {
+            statement.setLong(1, car.getManufacturer().getId());
+            statement.setString(2, car.getModel());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                car.setId(resultSet.getObject("id",Long.class));
+            }
+            if (!car.getDrivers().isEmpty()) {
```
Comment (dmytro-hryhoruk) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: let's also check on null

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                statement.setLong(2, driver.getId());
+                statement.executeUpdate();
+            } catch (SQLException ex) {
+                throw new DataProcessingException("Add driver=" + driver + " to car=" + car
+                        + " is failed", ex);
+            }
+        }
+    }
+
+    private Car getCarFromResultSet(Connection connection, ResultSet resultSet)
```
Comment (dmytro-hryhoruk) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: is this the only thing this method does?

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        String query = "SELECT c.id, c.model, m.id AS m_id, m.name, m.country "
+                + "FROM cars c "
+                + "LEFT JOIN cars_drivers cd on c.id = cd.car_id "
+                + "LEFT JOIN drivers d on cd.driver_id = d.id "
+                + "LEFT JOIN manufacturers m on m.id = c.m_id "
+                + "WHERE d.id = ? AND c.is_deleted = false;";
```
Comment (dmytro-hryhoruk) [inline src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: check cars on deleted

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 3 above)
Comment (dmytro-hryhoruk) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: this won't have any effect, let's remove this. You are updating BY id, so id remains the same.

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 7 above)
Comment (Vedroid) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: @dmytro-hryhoruk Isn't that it?:  `AND c.is_deleted = false;`

File: src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java (hunk 6 above)
Comment (Vedroid) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: Yes, but i can rename to `getCarWithManufacturerAndDriversFromResultSet()`

Comment (dmytro-hryhoruk) [reply src/main/java/ua/vedroid/dao3/dao/jdbc/CarDaoJdbcImpl.java]: @Vedroid thisMethod not only parsing resultSet, but also making additional request to fetch all drivers and also to parse them too. Let's divide logic of parsing result set for car and of fetching drivers for this car.
you will have two private methods:
getCarFromResultSet()
getDrivers()
and use them both in places where you were using getCarFromResultSet()

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**